zkvm/src/vm.rs

```diff
@@ -586,90 +599,26 @@ where
-        let utxo = UTXO::from_output(output_slice, &txid);
-        Ok((contract, utxo))
-    }
-
-    fn decode_output<'a>(&mut self, mut output: Subslice<'a>) -> Result<Contract, VMError> {
-        //    Output  =  Predicate  ||  LE32(k)  ||  Item[0]  || ... ||  Item[k-1]
-        // Predicate  =  <32 bytes>
-        //      Item  =  enum { Data, Value }
-        //      Data  =  0x00  ||  LE32(len)  ||  <bytes>
-        //     Value  =  0x01  ||  <32 bytes> ||  <32 bytes>
-
-        let predicate = Predicate::opaque(output.read_point()?);
-        let k = output.read_size()?;
-
-        // sanity check: avoid allocating unreasonably more memory
-        // just because an untrusted length prefix says so.
-        if k > output.len() {
-            return Err(VMError::FormatError);
-        }
-
-        let mut payload: Vec<PortableItem> = Vec::with_capacity(k);
-        for _ in 0..k {
-            let item = match output.read_u8()? {
-                DATA_TYPE => {
-                    let len = output.read_size()?;
-                    let bytes = output.read_bytes(len)?;
-                    PortableItem::Data(Data::Opaque(bytes.to_vec()))
-                }
-                VALUE_TYPE => {
-                    let qty = output.read_point()?;
-                    let flv = output.read_point()?;
-
-                    let qty = self.make_variable(Commitment::Closed(qty));
-                    let flv = self.make_variable(Commitment::Closed(flv));
-
-                    PortableItem::Value(Value { qty, flv })
-                }
-                _ => return Err(VMError::FormatError),
-            };
-            payload.push(item);
-        }
-
-        Ok(Contract { predicate, payload })
-    }
-
-    fn encode_output(&mut self, contract: Contract) -> Vec<u8> {
-        let mut output = Vec::with_capacity(contract.min_serialized_length());
-
-        encoding::write_point(&contract.predicate.point(), &mut output);
-        encoding::write_u32(contract.payload.len() as u32, &mut output);
-
-        for item in contract.payload.iter() {
-            match item {
-                PortableItem::Data(d) => match d {
-                    Data::Opaque(data) => {
-                        encoding::write_u8(DATA_TYPE, &mut output);
-                        encoding::write_u32(data.len() as u32, &mut output);
-                        encoding::write_bytes(&data, &mut output);
-                    }
-                    Data::Witness(_) => unimplemented!(),
-                },
-                PortableItem::Value(v) => {
-                    encoding::write_u8(VALUE_TYPE, &mut output);
-                    let qty = self.get_variable_commitment(v.qty);
-                    let flv = self.get_variable_commitment(v.flv);
-                    encoding::write_point(&qty, &mut output);
-                    encoding::write_point(&flv, &mut output);
-                }
-            }
+        let contract = self.unfreeze_contract(input.contract);
```

Comment:
Maybe to make it symmetrical with `output`, we can inline this logic inside `input`? So there's only one helper method - for freezing and unfreezing.

```rust
    fn input(&mut self) -> Result<(), VMError> {
        let input = self.pop_item()?.to_data()?.to_input()?;
        let contract = self.unfreeze_contract(input.contract);
        self.push_item(contract);
        self.txlog.push(Entry::Input(input.utxo));
        self.unique = true;
        Ok(())
    }
```